torch/onnx/README.md

```diff
@@ -23,3 +24,75 @@ symbolic_opset9.py.
+- Parameter names must *exactly* match the names in
+  aten/src/ATen/native/native_functions.yaml, because
+  dispatch is done with keyword arguments.
+- Looking for inplace ops? They're detected by
+  `_jit_pass_onnx_remove_inplace_ops_for_onnx`, and
+  transparently dispatched to their non inplace versions in
+  "run_symbolic_function". See Note [Export inplace](#export-inplace)
+- Required: Annotate new symbolic functions with type annotations and decorate with `@_beartype.beartype` to enable runtime type checking.
+
+### A note on Tensor types
+
+In general, we should avoid depending on the type of Tensor Values contained
+within the trace graph. However, this is sometimes unavoidable (due to ONNX
+spec requirements, etc). The TensorType object has accessors for these properties that return the property if it is statically known and return nullopt otherwise.
+
+In general, we should prefer to rely on the least specific information possible.
+For example, not relying on tensor properties at all is better than relying
+on the number of dimensions which is better than relying on
+concrete shapes. Doing so will make the export symbolics
+more robust to different graphs.
+
+### Extra context for symbolic functions
+
+> TODO(justinchuby): Update this after #84776 is merged.
+
+In general, symbolic functions only require inputs and attributes to
+the original node. In rare circumstances, extra context may be required.
+For example, symbolic function for `prim::Loop` needs access to the sub-block of
+the original node.
+A symbolic function that has a first arg (before the Graph object) with the
+type annotation of torch.onnx.SymbolicContext will be called with that additional context.
+During export, it is populated from `utils._run_symbolic_function`
+to contain the context for each node being converted.
+
+### Export inplace
+
+It would be better for us to export inplace annotations,
+than to not export them, since it is useful information that can
+help the target of an ONNX export export more efficiently. However,
+ONNX doesn't currently formalize inplace. Fortunately, it's sound to drop
+inplace annotations, but we are losing information this way.
+
+### Pointwise by scalar
+
+What happens if you add a tensor with a constant (e.g., x + 2)?  There are
+some moving parts to implementing the ONNX translation in this case:
+
+- By the time we get the scalar in a symbolic function here, it is no longer
+    a Python long/float, but a PyTorch tensor with `numel == 1` (eventually, we
+    want it to be a zero dim tensor but this change has not happened yet.)
+    However, the type of this scalar is *exactly* what the user wrote in
+    Python, which may not match the tensor it is being added to.  PyTorch
+    will do implicit conversions on scalars; however, ONNX will not, so
+    we must do the conversion ourselves.  This is what `symbolic_helper._if_scalar_type_as()`
+    does.
```

Comment:
Noticed this is a little outdated, best add
```suggestion
    and `_jit_pass_onnx_scalar_type_analysis` does.
```

Reply:
Done